liblouis/lou_translateString.c

```diff
@@ -3223,119 +3154,91 @@ beginCount(const EmphasisInfo *buffer, const int at, const EmphasisClass class,
 					input, output, posMapping, cursorPosition, cursorStatus);
+
+	if (end && other) {
+
+		/* end bits */
+		for (i = 0; i < 10; i++)
+			type_counts[i] = endCount(emphasisBuffer, at, emphClasses[i]);
+
+		for (i = 0; i < 10; i++) {
+			min = -1;
+			for (j = 0; j < 10; j++)
+				if (type_counts[j] > 0)
+					if (min < 0 || type_counts[j] < type_counts[min]) min = j;
+			if (min < 0) break;
+			type_counts[min] = 0;
+			insertEmphasisEnd(emphasisBuffer, at, emph1Rule + min, emphClasses[min],
+					table, pos, input, output, posMapping, cursorPosition, cursorStatus);
+		}
 	}
 
-	/* insert capitalization last so it will be closest to word */
-	if ((emphasisBuffer[at].begin | emphasisBuffer[at].end | emphasisBuffer[at].word |
-				emphasisBuffer[at].symbol) &
-			capsEmphClass) {
-		insertEmphasisBegin(emphasisBuffer, at, capsRule, capsEmphClass, table, pos,
-				input, output, posMapping, cursorPosition, cursorStatus);
-		insertEmphasisSymbol(emphasisBuffer, at, capsRule, capsEmphClass, table, pos,
-				input, output, posMapping, cursorPosition, cursorStatus);
+	if (begin && other) {
+
+		/* begin and word bits */
+		for (i = 0; i < 10; i++)
+			type_counts[i] = beginCount(emphasisBuffer, at, emphClasses[i], table, input);
+
+		for (i = 9; i >= 0; i--) {
+			max = 9;
+			for (j = 9; j >= 0; j--)
+				if (type_counts[max] < type_counts[j]) max = j;
+			if (!type_counts[max]) break;
+			type_counts[max] = 0;
+			insertEmphasisBegin(emphasisBuffer, at, emph1Rule + max, emphClasses[max],
+					table, pos, input, output, posMapping, cursorPosition, cursorStatus);
+		}
+
+		/* symbol bits */
+		for (i = 9; i >= 0; i--)
+			if ((emphasisBuffer[at].begin | emphasisBuffer[at].end |
+						emphasisBuffer[at].word | emphasisBuffer[at].symbol) &
+					emphClasses[i])
+				insertEmphasisSymbol(emphasisBuffer, at, emph1Rule + i, emphClasses[i],
+						table, pos, input, output, posMapping, cursorPosition,
+						cursorStatus);
 	}
-}
 
-static void
-insertEmphases(const TranslationTableHeader *table, int from, int to,
-		const InString *input, OutString *output, int *posMapping,
-		const EmphasisInfo *emphasisBuffer, int haveEmphasis, int transOpcode,
-		int *cursorPosition, int *cursorStatus) {
-	int at;
-	for (at = from; at <= to; at++)
-		insertEmphasesAt(at, table, to, input, output, posMapping, emphasisBuffer,
-				haveEmphasis, transOpcode, cursorPosition, cursorStatus);
+	if (begin && caps) {
+
+		/* insert capitalization last so it will be closest to word */
+		if ((emphasisBuffer[at].begin | emphasisBuffer[at].end | emphasisBuffer[at].word |
+					emphasisBuffer[at].symbol) &
+				capsEmphClass) {
+			insertEmphasisBegin(emphasisBuffer, at, capsRule, capsEmphClass, table, pos,
+					input, output, posMapping, cursorPosition, cursorStatus);
+			insertEmphasisSymbol(emphasisBuffer, at, capsRule, capsEmphClass, table, pos,
+					input, output, posMapping, cursorPosition, cursorStatus);
+		}
+	}
```

Comment:
In theory you'd need an else clause here that checks for invalid cases, e.g. `begin & end`

Reply:
There are no invalid cases.